commons/src/build_output.rs

```diff
@@ -0,0 +1,709 @@
+        let miliseconds = milliseconds(duration);
+
+        if hours > 0 {
+            format!("{hours}h {minutes}m {seconds}s")
+        } else if minutes > 0 {
+            format!("{minutes}m {seconds}s")
+        } else if seconds > 0 || miliseconds > 100 {
+            // 0.1
+            format!("{seconds}.{miliseconds:0>3}s")
+        } else {
+            String::from("< 0.1s")
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        #[test]
+        fn test_millis_and_seconds() {
+            let duration = Duration::from_millis(1024);
+            assert_eq!(24, milliseconds(&duration));
+            assert_eq!(1, seconds(&duration));
+        }
+
+        #[test]
+        fn test_display_duration() {
+            let duration = Duration::from_millis(99);
+            assert_eq!("< 0.1s", human(&duration).as_str());
+
+            let duration = Duration::from_millis(1024);
+            assert_eq!("1.024s", human(&duration).as_str());
+
+            let duration = std::time::Duration::from_millis(60 * 1024);
+            assert_eq!("1m 1s", human(&duration).as_str());
+
+            let duration = std::time::Duration::from_millis(3600 * 1024);
+            assert_eq!("1h 1m 26s", human(&duration).as_str());
+        }
+    }
+}
+
+// Helper for printing without newlines that auto-flushes stdout
+fn raw_inline_print(contents: impl AsRef<str>) {
+    let contents = contents.as_ref();
+    print!("{contents}");
+    std::io::stdout().flush().expect("Stdout is writable");
+}
+
+/// All work is done inside of a section. Advertize a section topic
+pub fn section(topic: impl AsRef<str>) -> section::Section {
+    let topic = String::from(topic.as_ref());
+    println!("- {topic}");
+
+    section::Section { topic }
+}
+
+/// Top level buildpack header
+///
+/// Should only use once per buildpack
+#[must_use]
+pub fn buildpack_name(buildpack: impl AsRef<str>) -> BuildpackDuration {
+    let header = fmt::header(buildpack.as_ref());
+    println!("{header}");
+    println!();
+
+    let start = Instant::now();
+    BuildpackDuration { start }
+}
+
+mod section {
+    use super::{
+        add_prefix, fmt, fun_run, line_mapped, raw_inline_print, time, time::LiveTimingInline,
+        CmdError, Instant,
+    };
+    use libherokubuildpack::command::CommandExt;
+    use std::process::{Command, Output};
+
+    const CMD_INDENT: &str = "      ";
+    const SECTION_INDENT: &str = "  ";
+    const SECTION_PREFIX: &str = "  - ";
+
+    #[derive(Debug, Clone, Eq, PartialEq)]
+    pub struct Section {
+        pub(crate) topic: String,
+    }
+
+    impl Section {
+        /// Emit contents to the buid output with indentation
+        pub fn say(&self, contents: impl AsRef<str>) {
+            let contents = contents.as_ref();
+            println!("{SECTION_PREFIX}{contents}");
+        }
+
+        pub fn say_with_details(&self, contents: impl AsRef<str>, details: impl AsRef<str>) {
```

Comment:
Given details are just messages wrapped in parentheses which are super easy for authors to use, I'm not sure this has a lot of value. If I want to put details mid-sentence, then I'd have to drop down to using fmt::Display or `(` literals myself. Honestly, I'd probably just use literals.

Reply:
This method is an extraction from a pattern I was doing over and over. You can manually format it with details, but it's a bunch of boilerplate:

```rust
let details = build_output::fmt::details("bundler version changed");
self.build_output
    .say(format!("Clearing cache {details}");
```

Versus this method needs no `format!` invocation and takes 2 lines instead of 3 (I also think it reads better):

```rust
self.build_output
    .say_with_details("Clearing cache", "bundler version changed");
```